MiniApp/Classes/ui/MiniAppViewController.swift

```diff
@@ -0,0 +1,193 @@
+            }
+            if navigationItem.rightBarButtonItems == nil {
+                navigationItem.setRightBarButton(closeButton, animated: true)
+            }
+        }
+
+        view.addSubview(activityIndicatorView)
+        activityIndicatorView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            activityIndicatorView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            activityIndicatorView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
+        ])
+
+        view.addSubview(fallbackView)
+        fallbackView.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            fallbackView.topAnchor.constraint(equalTo: view.topAnchor),
+            fallbackView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            fallbackView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            fallbackView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        ])
+    }
+
+    func setupMiniApp() {
+
+        state = .loading
+
+        guard let messageDelegate = messageDelegate else { return }
+        MiniApp.shared(with: config).create(
+            appId: appId,
+            queryParams: queryParams,
+            completionHandler: { [weak self] (result) in
+                guard let self = self else { return }
+                switch result {
+                case .success(let miniAppDisplay):
+                    let view = miniAppDisplay.getMiniAppView()
+                    view.frame = self.view.bounds
+                    self.view.addSubview(view)
+                    self.navBarDelegate = miniAppDisplay as? MiniAppNavigationBarDelegate
+                    self.delegate?.miniApp(self, didLoadWith: nil)
+                    self.state = .success
+                case .failure(let error):
+                    self.delegate?.miniApp(self, didLoadWith: error)
+                    self.state = .error
+                }
+            },
+            messageInterface: messageDelegate
+        )
+    }
+
+    func setup() {
+        fallbackView.onRetry = { [weak self] in
+            guard let self = self else { return }
+            self.setupMiniApp()
+        }
+    }
+
+    func update() {
+        DispatchQueue.main.async {
+            switch self.state {
+            case .loading:
+                self.activityIndicatorView.startAnimating()
+                self.fallbackView.isHidden = true
+            case .error:
+                self.activityIndicatorView.stopAnimating()
+                self.fallbackView.isHidden = false
+            case .success:
+                self.activityIndicatorView.stopAnimating()
+                self.fallbackView.isHidden = true
+            }
+        }
+    }
+
+    @objc
+    public func backPressed() {
+        if delegate == nil {
+            delegate?.miniApp(self, shouldExecute: .back)
```

Comment:
🤔 . Shouldn't this be !=

Reply:
Good catch @ts-leojoseph-r might be right ^^. Maybe a `if let delegate = delegate` would remove the need to unwrap it :) I don't know what's the most elegant to be honest.
I'm wondering though, shouldn't we call both delegates if they are present?

Reply:
ouh woops 😂